database/utxo.go

```diff
@@ -116,104 +117,71 @@ func (d *Database) UtxosByAddress(
 	return ret, nil
 }
 
-func UtxoDelete(
+func UtxosCleanup(
 	db *Database,
-	utxo Utxo,
+	slot uint64,
 ) error {
-	return db.UtxoDelete(utxo, nil)
+	return db.UtxosCleanup(slot, nil)
 }
 
-func (d *Database) UtxoDelete(
-	utxo Utxo,
+func (d *Database) UtxosCleanup(
+	slot uint64,
 	txn *Txn,
 ) error {
+	var ret error
 	if txn == nil {
-		txn = d.Transaction(true)
+		txn = d.Transaction(false)
 		defer txn.Commit() //nolint:errcheck
 	}
-	// Remove from metadata DB
-	err := d.metadata.DeleteUtxo(utxo, txn.Metadata())
-	if err != nil {
-		return err
-	}
-	// Remove from blob DB
-	key := UtxoBlobKey(utxo.TxId, utxo.OutputIdx)
-	err = txn.Blob().Delete(key)
+	// Get UTxOs that are marked as deleted and older than our slot window
+	utxos, err := d.metadata.GetUtxosDeletedBeforeSlot(slot, txn.Metadata())
 	if err != nil {
-		return err
+		return errors.New("failed to query consumed UTxOs during cleanup")
 	}
-	return nil
-}
-
-func UtxosDelete(
-	db *Database,
-	utxos []Utxo,
-) error {
-	return db.UtxosDelete(utxos, nil)
-}
 
-func (d *Database) UtxosDelete(
-	utxos []Utxo,
-	txn *Txn,
-) error {
-	if txn == nil {
-		txn = d.Transaction(true)
-		defer txn.Commit() //nolint:errcheck
-	}
-	// Remove from metadata DB
-	tmpUtxos := []any{}
-	for _, utxo := range utxos {
-		tmpUtxos = append(tmpUtxos, utxo)
-	}
-	err := d.metadata.DeleteUtxos(tmpUtxos, txn.Metadata())
-	if err != nil {
-		return err
-	}
-	// Remove from blob DB
-	for _, utxo := range utxos {
-		key := UtxoBlobKey(utxo.TxId, utxo.OutputIdx)
-		err := txn.Blob().Delete(key)
+	// Loop through UTxOs and delete, with a new transaction each loop
+	for {
+		// short-circuit loop
+		if ret != nil {
+			break
+		}
+		batchSize := min(1000, len(utxos))
+		if batchSize == 0 {
+			break
+		}
+		// Delete the UTxOs
+		loopTxn := d.Transaction(true)
+		// Remove from metadata DB
+		tmpUtxos := []any{}
+		for _, utxo := range utxos[0:batchSize] {
+			tmpUtxos = append(tmpUtxos, utxo)
+		}
+		err := d.metadata.DeleteUtxos(tmpUtxos, loopTxn.Metadata())
 		if err != nil {
-			return err
+			ret = err
+			break
 		}
-	}
-	return nil
-}
-
-func UtxosDeleted(
-	db *Database,
-	slot uint64,
-) ([]Utxo, error) {
-	return db.UtxosDeleted(slot, nil)
-}
-
-func (d *Database) UtxosDeleted(
-	slot uint64,
-	txn *Txn,
-) ([]Utxo, error) {
-	ret := []Utxo{}
-	if txn == nil {
-		txn = d.Transaction(false)
-		defer txn.Commit() //nolint:errcheck
-	}
-	utxos, err := d.metadata.GetUtxosDeletedBeforeSlot(slot, txn.Metadata())
-	if err != nil {
-		return ret, err
-	}
-	for _, utxo := range utxos {
-		tmpUtxo := Utxo(utxo)
-		if err := tmpUtxo.loadCbor(txn); err != nil {
-			return ret, err
+		// Remove from blob DB
+		for _, utxo := range utxos[0:batchSize] {
+			key := UtxoBlobKey(utxo.TxId, utxo.OutputIdx)
+			err := loopTxn.Blob().Delete(key)
+			if err != nil {
+				ret = err
+				break
+			}
 		}
-		ret = append(ret, tmpUtxo)
+		loopTxn.Commit()
```

Comment:
Instead of managing the DB txn directly, you can wrap everything between here and where the txn is created above in a `txn.Do(...)`, which will manage closing the txn itself, even on error